src/core/cff_font.js

```diff
@@ -48,26 +48,42 @@ class CFFFont {
   getGlyphMapping() {
     const cff = this.cff;
     const properties = this.properties;
+    const { cidToGidMap, cMap } = properties;
     const charsets = cff.charset.charset;
     let charCodeToGlyphId;
     let glyphId;
 
     if (properties.composite) {
+      let invCidToGidMap;
+      if (cidToGidMap && cidToGidMap.length > 0) {
+        invCidToGidMap = Object.create(null);
+        for (let i = 0, ii = cidToGidMap.length; i < ii; i++) {
+          const gid = cidToGidMap[i];
+          if (gid !== undefined) {
+            invCidToGidMap[gid] = i;
+          }
+        }
+      }
+
       charCodeToGlyphId = Object.create(null);
       let charCode;
       if (cff.isCIDFont) {
         // If the font is actually a CID font then we should use the charset
         // to map CIDs to GIDs.
         for (glyphId = 0; glyphId < charsets.length; glyphId++) {
           const cid = charsets[glyphId];
-          charCode = properties.cMap.charCodeOf(cid);
+          charCode = cMap.charCodeOf(cid);
+
+          if (invCidToGidMap && invCidToGidMap[charCode] !== undefined) {
+            charCode = invCidToGidMap[charCode];
+          }
```

Comment:
As mentioned in https://github.com/mozilla/pdf.js/pull/15563#issue-1405057918, I don't really understand this well enough to know how it *should* work.
However this patch is the only way that I could come that actually fixes the issue, since trying to do apply the /CIDToGIDMap elsewhere in this function doesn't actually fix the issue :-(

Reply:
We have (25573 is the char code for the first ideograph just after FutureABC L3):
 - charsets: 25573 (gid) => 40307 (cid)
 - cidToGidMap: 25573 (cid) => 40307 (gid)
 - charsets: 17688 (gid) => 25573 (cid)

If I remove the `/CidToGidMap` entry for the font in the pdf then the rendering is incorrect in either Chrome and Acrobat so this map really matters.
A possibility is that the gid in `charsets` is a detail of implementation and the exposed id is finally the cid from charsets which could be a kind of `gid`, hence `cidToGidMap` would map from a cid in the pdf to the kind of gid in the font.
I'm loosing myself when I read the above sentence...
Anyway, your patch could make sense.
I'm not sure if we should apply `invCidToGid` to `cid` or to `charCode`, wdyt ?

Reply:
And it's interesting to read pdium code:
https://pdfium.googlesource.com/pdfium/+/refs/heads/main/core/fpdfapi/font/cpdf_cidfont.cpp#689

and the value from `CIDToGIDMap` is returned here: https://pdfium.googlesource.com/pdfium/+/refs/heads/main/core/fpdfapi/font/cpdf_cidfont.cpp#824
and it seems to be the `gid` and the only way to get the correct final glyph id is to reverse the value from charsets.

Reply:
And we should really add some comments to explain why we're doing that.

Reply:
> I'm loosing myself when I read the above sentence...

Yeah, all of this is really difficult to get your head around. Looking through both the PDF and the CFF specifications doesn't really help either.
According to the [relevant section in the PDF spec](https://web.archive.org/web/20220309040754if_/https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf#page=278) it even sounds like these fonts shouldn't be using a /CIDToGIDMap map (but it wouldn't be the first time that the PDF spec is incomplete/incorrect), since the relevant fonts are actually CFF and not TrueType (excerpt below):
```
This entry may appear only in a Type 2 CIDFont whose associated
TrueType font program is embedded in the PDF file.
```

> I'm not sure if we should apply `invCidToGid` to `cid` or to `charCode`, wdyt ?

Based on *quick* testing, changing that doesn't seem to work unfortunately.

> And we should really add some comments to explain why we're doing that.

I'm really struggling to explain what's happening here, do you have any suggestions as to its wording?

Reply:
> 
> I'm really struggling to explain what's happening here, do you have any suggestions as to its wording?

If only I was able to....
Maybe just copy/paste the example I put in https://github.com/mozilla/pdf.js/pull/15563#discussion_r993422718
and something like "it seems that the GID in CIDToGIDMap corresponds to the CID in the CFF font".

Reply:
I've tried to add a brief comment, and hopefully anyone curious about further details can reference this issue via `blame` (adding details from a specific font *and* glyph to the code doesn't really feel all that much clearer, at least to me).